airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java

```diff
@@ -197,31 +206,55 @@ private void produceMessages(final AutoCloseableIterator<AirbyteMessage> message
   }
 
   @VisibleForTesting
-  static void consumeWriteStream(final AirbyteMessageConsumer consumer) throws Exception {
-    // use a Scanner that only processes new line characters to strictly abide with the
-    // https://jsonlines.org/ standard
-    final Scanner input = new Scanner(System.in, StandardCharsets.UTF_8).useDelimiter("[\r\n]+");
-    consumer.start();
-    while (input.hasNext()) {
-      consumeMessage(consumer, input.next());
+  static void consumeWriteStream(final SerializedAirbyteMessageConsumer consumer) throws Exception {
+    try (final BufferedInputStream bis = new BufferedInputStream(System.in);
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      consumeWriteStream(consumer, bis, baos);
     }
   }
 
-  private static void runConsumer(final AirbyteMessageConsumer consumer) throws Exception {
-    watchForOrphanThreads(
-        () -> consumeWriteStream(consumer),
-        () -> System.exit(FORCED_EXIT_CODE),
-        INTERRUPT_THREAD_DELAY_MINUTES,
-        TimeUnit.MINUTES,
-        EXIT_THREAD_DELAY_MINUTES,
-        TimeUnit.MINUTES);
+  @VisibleForTesting
+  static void consumeWriteStream(final SerializedAirbyteMessageConsumer consumer,
+                                 final BufferedInputStream bis,
+                                 final ByteArrayOutputStream baos)
+      throws Exception {
+    consumer.start();
+
+    final byte[] buffer = new byte[8192]; // 8K buffer
+    int bytesRead;
+    int byteCount = 0;
+    boolean lastWasNewLine = false;
+
+    while ((bytesRead = bis.read(buffer)) != -1) {
```

Comment:
this looks functionally equivalent to scanner.nextLine. Is there a performance advantage to doing it manually?

I'd prefer to either explore a prebuilt class or really beef up the tests in IntegrationRunnerBackwardsCompatbilityTest. E.g. verify behavior with very large lines (>8192 bytes), multibyte characters, etc. Possibly even generating some randomized inputs / fuzzing. This sort of low-level byte handling makes me very nervous.

(SO also suggests using `BufferedReader#readLine` https://stackoverflow.com/a/5172393)

Reply:
we want to count the number of bytes we have per message. i went with the lower-level thing so that we could count the bytes as we pull them off the wire. otherwise we have to call `getBytes("UTF")).size()` on the string which isn't free. admittedly, i haven't tested the performance difference, but looking at the impl it looks like it has to allocate a new byte array to do this.

Reply:
how about something like this? it doesn't exclude newline/carriage return characters, but presumably that's a negligible difference
```java
try (CountingInputStream counter = new CountingInputStream(System.in);
    BufferedReader reader = new BufferedReader(new InputStreamReader(counter))) {
  while (true) {
    String line = reader.readLine();
    if (line == null) {
      break;
    }
    long messageBytes = counter.getCount();
    counter.reset();
  }
}
```

Reply:
🔥 🔥 🔥  i didn't know this was a thing. let me try it! if it works, i agree this is way better.

Reply:
womp. I might mess around with this later today. I feel like there's some ordering of reader -> counter -> buffer that could work. agree with merging this pr though.